@Step doesn't work in Before/After

#### I'm submitting a ... 
  - [x] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?

AllureLifecycle is failing in Before/After methods
All attachments in steps also are lost ( because there are no Steps, your Cap :) )

```
07:59:26.911 [main] ERROR io.qameta.allure.AllureLifecycle - Could not add attachment: no test is running

07:59:26.912 [main] ERROR io.qameta.allure.AllureLifecycle - Could not update step: step with uuid c3cb22ea-18a5-4083-8cf3-79c0070997aa not found

07:59:26.912 [main] ERROR io.qameta.allure.AllureLifecycle - Could not stop step: step with uuid c3cb22ea-18a5-4083-8cf3-79c0070997aa not found
```



#### If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem


#### What is the expected behavior?

Steps added in Before/After methods

#### What is the motivation / use case for changing the behavior?

Make Allure moar great

#### Please tell us about your environment:


| Allure version | 2.8.1 |
| --- | --- |
| Test framework | testng@6.14.3 |
| Allure integration | allure-testng@2.8.1 |
| Generate report using | allure-maven@2.8.1|

#### Other information 

[//]: # (
. e.g. detailed explanation, stacktraces, related issues, suggestions 
. how to fix, links for us to have more context, eg. Stackoverflow, Gitter etc
)

@baev use this example for reproduce issue
`public class IssueTest {

    @BeforeClass
    public static void checkForData() {
        beforeMethod();
    }

    @Step("Step before")
    private static void beforeMethod() {
        System.out.println("before");
    }

    @Step("Step in dp")
    protected static Object createDpData() {
        System.out.println("dp step");
        return new Object();
    }

    @DataProvider(name = "dataprovider")
    public Object[][] data() {
        List<Object[]> data = new ArrayList<>();
        Object p = createDpData();
        data.add(new Object[]{p});
        return data.toArray(new Object[data.size()][]);
    }

    @Test(dataProvider = "dataprovider")
    public void test(final Object obj) {
        System.out.println("test");
    }
}`
u'll see this stdout
`before
2018-12-24 16:36:30[ERROR] Could not start step: no test case running
dp step
2018-12-24 16:36:30[ERROR] Could not update step: step with uuid 35e641d7-76cf-4d62-ad17-7811b21a89d4 not found
2018-12-24 16:36:30[ERROR] Could not stop step: step with uuid 35e641d7-76cf-4d62-ad17-7811b21a89d4 not found
test`

testNG 6.14.3
allure testNG 2.8.1
allure 2.7.0
allure maven 2.9.0

@dimkin-eu @baev trying to reproduce issue with junit 4.12 =>
`public class IssueTest {

    @Test
    @DisplayName("Not working")
    public void test() {
        System.out.println("start");
        step();
        System.out.println("finish");
    }


    @Step("Step")
    public void step() {
        System.out.println("step");
    }
}`

1) If i run this test from maven test -Dtest=IssueTest => all good
2) If i run this test from intellijiIdea => i have errors from allure - 
`test
2018-12-26 13:39:51[ERROR] Could not start step: no test case running
step
2018-12-26 13:39:51[ERROR] Could not update step: step with uuid 8dad6a72-736d-48d5-9fae-3673eb822d85 not found
2018-12-26 13:39:51[ERROR] Could not stop step: step with uuid 8dad6a72-736d-48d5-9fae-3673eb822d85 not found
done`

--- a/allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java
+++ b/allure-testng/src/main/java/io/qameta/allure/testng/AllureTestNg.java
@@ -61,6 +61,8 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -106,37 +108,26 @@ public class AllureTestNg implements
     /**
      * Store current testng result uuid to attach before/after methods into.
      */
-    private final ThreadLocal<Current> currentTestResult = new InheritableThreadLocal<Current>() {
-        @Override
-        protected Current initialValue() {
-            return new Current();
-        }
-    };
+    private final ThreadLocal<Current> currentTestResult = ThreadLocal
+            .withInitial(Current::new);
 
     /**
      * Store current container uuid for fake containers around before/after methods.
      */
-    private final ThreadLocal<String> currentTestContainer = new InheritableThreadLocal<String>() {
-        @Override
-        protected String initialValue() {
-            return UUID.randomUUID().toString();
-        }
-    };
+    private final ThreadLocal<String> currentTestContainer = ThreadLocal
+            .withInitial(() -> UUID.randomUUID().toString());
 
     /**
      * Store uuid for current executable item to catch steps and attachments.
      */
-    private final ThreadLocal<String> currentExecutable = new InheritableThreadLocal<String>() {
-        @Override
-        protected String initialValue() {
-            return UUID.randomUUID().toString();
-        }
-    };
+    private final ThreadLocal<String> currentExecutable = ThreadLocal
+            .withInitial(() -> UUID.randomUUID().toString());
 
     /**
      * Store uuid for class test containers.
      */
     private final Map<ITestClass, String> classContainerUuidStorage = new ConcurrentHashMap<>();
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
 
     private final AllureLifecycle lifecycle;
 
@@ -215,16 +206,14 @@ public class AllureTestNg implements
                 .setUuid(uuid)
                 .setName(testClass.getName());
         getLifecycle().startTestContainer(container);
-        classContainerUuidStorage.put(testClass, uuid);
+        setClassContainer(testClass, uuid);
     }
 
     public void onAfterClass(final ITestClass testClass) {
-        if (!classContainerUuidStorage.containsKey(testClass)) {
-            return;
-        }
-        final String uuid = classContainerUuidStorage.get(testClass);
-        getLifecycle().stopTestContainer(uuid);
-        getLifecycle().writeTestContainer(uuid);
+        getClassContainer(testClass).ifPresent(uuid -> {
+            getLifecycle().stopTestContainer(uuid);
+            getLifecycle().writeTestContainer(uuid);
+        });
     }
 
     @Override
@@ -242,11 +231,7 @@ public class AllureTestNg implements
         Optional.of(testResult)
                 .map(ITestResult::getMethod)
                 .map(ITestNGMethod::getTestClass)
-                .map(classContainerUuidStorage::get)
-                .ifPresent(testClassContainerUuid -> getLifecycle().updateTestContainer(
-                        testClassContainerUuid,
-                        container -> container.getChildren().add(uuid)
-                ));
+                .ifPresent(clazz -> addClassContainerChild(clazz, uuid));
     }
 
     protected void startTestCase(final ITestResult testResult,
@@ -400,12 +385,12 @@ public class AllureTestNg implements
 
     private void ifClassFixtureStarted(final ITestNGMethod testMethod) {
         if (testMethod.isBeforeClassConfiguration()) {
-            final String parentUuid = classContainerUuidStorage.get(testMethod.getTestClass());
-            startBefore(parentUuid, testMethod);
+            getClassContainer(testMethod.getTestClass())
+                    .ifPresent(parentUuid -> startBefore(parentUuid, testMethod));
         }
         if (testMethod.isAfterClassConfiguration()) {
-            final String parentUuid = classContainerUuidStorage.get(testMethod.getTestClass());
-            startAfter(parentUuid, testMethod);
+            getClassContainer(testMethod.getTestClass())
+                    .ifPresent(parentUuid -> startAfter(parentUuid, testMethod));
         }
     }
 
@@ -694,6 +679,39 @@ public class AllureTestNg implements
                 result.getStatusDetails().setMessage(details.getMessage());
             }
         };
+    }
+
+    private void addClassContainerChild(final ITestClass clazz, final String childUuid) {
+        lock.writeLock().lock();
+        try {
+            final String parentUuid = classContainerUuidStorage.get(clazz);
+            if (nonNull(parentUuid)) {
+                getLifecycle().updateTestContainer(
+                        parentUuid,
+                        container -> container.getChildren().add(childUuid)
+                );
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private Optional<String> getClassContainer(final ITestClass clazz) {
+        lock.readLock().lock();
+        try {
+            return Optional.ofNullable(classContainerUuidStorage.get(clazz));
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    private void setClassContainer(final ITestClass clazz, final String uuid) {
+        lock.writeLock().lock();
+        try {
+            classContainerUuidStorage.put(clazz, uuid);
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
     private Current refreshContext() {

--- a/allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java
+++ b/allure-testng/src/test/java/io/qameta/allure/testng/AllureTestNgTest.java
@@ -335,7 +335,7 @@ public class AllureTestNgTest {
     }
 
     @AllureFeatures.Fixtures
-    @Test(description = "Class fixtures", dataProvider = "parallelConfiguration", enabled = false)
+    @Test(description = "Class fixtures", dataProvider = "parallelConfiguration")
     public void perClassFixtures(final XmlSuite.ParallelMode mode, final int threadCount) {
         final AllureResults results = runTestNgSuites(
                 parallel(mode, threadCount),
@@ -496,7 +496,7 @@ public class AllureTestNgTest {
     }
 
     @AllureFeatures.Parallel
-    @Test(description = "Parallel methods", enabled = false)
+    @Test(description = "Parallel methods")
     public void parallelMethods() {
         String before1 = "io.qameta.allure.testng.samples.ParallelMethods.beforeMethod";
         String before2 = "io.qameta.allure.testng.samples.ParallelMethods.beforeMethod2";
@@ -928,8 +928,7 @@ public class AllureTestNgTest {
     @Issue("219")
     @Test(
             description = "Should not mix up fixtures during parallel run",
-            dataProvider = "parallelConfiguration",
-            enabled = false
+            dataProvider = "parallelConfiguration"
     )
     public void shouldAddCorrectBeforeMethodFixturesInCaseOfParallelRun(
             final XmlSuite.ParallelMode mode, final int threadCount) {

--- a/allure-testng/src/test/java/io/qameta/allure/testng/samples/BeforeMethods.java
+++ b/allure-testng/src/test/java/io/qameta/allure/testng/samples/BeforeMethods.java
@@ -29,14 +29,6 @@ import org.testng.annotations.Test;
  */
 public class BeforeMethods {
 
-//    @BeforeSuite
-//    public void beforeSuite() {
-//    }
-//
-//    @AfterSuite
-//    public void afterSuite() {
-//    }
-
     @BeforeTest
     public void beforeTest() {
     }

